pkg/nad/masterplugin.go

```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
 	if ipam == nil {
 		glog.V(100).Infof("error adding empty ipam to MasterBridgePlugin")
 
-		plugin.errorMsg = "invalid ipam parameter"
+		plugin.errorMsg = invalidIpamParameterMsg
 	}
 
 	plugin.masterPlugin.Ipam = ipam
 
 	return plugin
 }
+
+// MasterVlanPlugin provides struct for MasterPlugin set to vlan in NetworkAttachmentDefinition.
+type MasterVlanPlugin struct {
+	masterPlugin *MasterPlugin
+	errorMsg     string
+}
+
+// NewMasterVlanPlugin creates new instance of MasterVlanPlugin.
+func NewMasterVlanPlugin(name string, vlanID uint16) *MasterVlanPlugin {
+	glog.V(100).Infof(
+		"Initializing new MasterVlanPlugin structure %s, with vlanId %s", vlanID)
+
+	builder := MasterVlanPlugin{
+		masterPlugin: &MasterPlugin{
+			CniVersion: "0.3.1",
+			Name:       name,
+			Type:       "vlan",
+			VlanID:     vlanID,
+		},
+	}
+
+	if vlanID > 4094 {
+		glog.V(100).Infof("error vlan id can not be greater than 4094")
+
+		builder.errorMsg = "MasterVlanPlugin vlanID is greater than 4094"
+	}
+
+	if builder.masterPlugin.Name == "" {
+		glog.V(100).Infof("error MasterVlanPlugin name can not be empty")
+
+		builder.errorMsg = "MasterVlanPlugin name is empty"
+	}
+
+	return &builder
+}
+
+// WithIPAM defines IPAM configuration to MasterVlanPlugin. Default is empty.
+func (plugin *MasterVlanPlugin) WithIPAM(ipam *IPAM) *MasterVlanPlugin {
+	glog.V(100).Infof("Adding IPAM configuration to MasterVlanPlugin: %v", ipam)
+
+	if plugin.masterPlugin == nil {
+		glog.V(100).Infof(msg.UndefinedCrdObjectErrString("MasterVlanPlugin"))
+		plugin.errorMsg = msg.UndefinedCrdObjectErrString("MasterVlanPlugin")
+	}
+
+	if ipam == nil {
+		glog.V(100).Infof("error adding empty ipam to MasterVlanPlugin")
+
+		plugin.errorMsg = invalidIpamParameterMsg
+	}
+
+	plugin.masterPlugin.Ipam = ipam
+
+	return plugin
+}
+
+// WithMasterInterface defines master interface to MasterVlanPlugin. Default is cn0.
+func (plugin *MasterVlanPlugin) WithMasterInterface(masterInterfaceName string) *MasterVlanPlugin {
+	glog.V(100).Infof("Adding masterInterfaceName interface %s to MasterVlanPlugin", masterInterfaceName)
+
+	if plugin.masterPlugin == nil {
+		glog.V(100).Infof(msg.UndefinedCrdObjectErrString("MasterVlanPlugin"))
+		plugin.errorMsg = msg.UndefinedCrdObjectErrString("MasterVlanPlugin")
+	}
+
+	if masterInterfaceName == "" {
+		glog.V(100).Infof("error to add masterInterfaceName interface, the name of interface can not be empty")
+
+		plugin.errorMsg = "invalid masterInterfaceName parameter"
+	}
```

Comment:
shouldn't you check if errorMsg != "" and return here?

Reply:
No I have this check before I return ready struct in GetMasterPluginConfig